source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -166,8 +159,15 @@ void AuthenticatorImpl::startVerify() {
   // Check the issuer is configured or not.
   jwks_data_ = provider_ ? jwks_cache_.findByProvider(provider_.value())
                          : jwks_cache_.findByIssuer(jwt_->iss_);
-  // isIssuerSpecified() check already make sure the issuer is in the cache.
-  ASSERT(jwks_data_ != nullptr);
+  // When `provider` is valid, findByProvider should never return nullptr.
+  // Only when `allow_missing` or `allow_failed` is used, `provider` is invalid,
```

Comment:
I'm fairly confused about this

> Only when `allow_missing` or `allow_failed` is used, `provider` is invalid

The code here suggests `provider_` can be set even when `allow_missing` is true.

https://github.com/envoyproxy/envoy/blob/fdd1693b49b5bf0d0cea9bed839c3b98f3b1b6e8/source/extensions/filters/http/jwt_authn/authenticator.cc#L105

Can you clarify why `provider_` should be nullptr?

Reply:
When allow_missing or allow_failed is used, the authenticator is designed to work on all providers, not just for one particular provider.  Hence provider is invalid.